commons-core/src/main/scala/com/avsystem/commons/serialization/json/JsonStringInput.scala

```diff
@@ -42,49 +47,55 @@ class JsonStringInput(reader: JsonReader, callback: AfterElement = AfterElementN
   }
 
   private def matchNumericString[T](toNumber: String => T): T = {
-    if (reader.jsonType == JsonType.number ||
-      (reader.jsonType == JsonType.string &&
-        (reader.currentValue == "Infinity" || reader.currentValue == "-Infinity" || reader.currentValue == "NaN"))) {
+    if (reader.jsonType == JsonType.number || reader.jsonType == JsonType.string) {
       val str = reader.currentValue.asInstanceOf[String]
       try toNumber(str) catch {
         case e: NumberFormatException => throw new ReadFailure(s"Invalid number format: $str", e)
       }
     } else expectedError(JsonType.number)
   }
 
-  def inputType: InputType = reader.jsonType match {
-    case JsonType.list => InputType.List
-    case JsonType.`object` => InputType.Object
-    case JsonType.`null` => InputType.Null
-    case _ => InputType.Simple
-  }
-
+  def isNull: Boolean = reader.jsonType == JsonType.`null`
   def readNull(): Null = checkedValue[Null](JsonType.`null`)
   def readString(): String = checkedValue[String](JsonType.string)
   def readBoolean(): Boolean = checkedValue[Boolean](JsonType.boolean)
   def readInt(): Int = matchNumericString(_.toInt)
   def readLong(): Long = matchNumericString(_.toLong)
   def readDouble(): Double = matchNumericString(_.toDouble)
   def readBigInt(): BigInt = matchNumericString(BigInt(_))
-  def readBigDecimal(): BigDecimal = matchNumericString(BigDecimal(_))
-  def readBinary(): Array[Byte] = {
-    val hex = checkedValue[String](JsonType.string)
-    val result = new Array[Byte](hex.length / 2)
-    var i = 0
-    while (i < result.length) {
-      result(i) = ((reader.fromHex(hex.charAt(2 * i)) << 4) | reader.fromHex(hex.charAt(2 * i + 1))).toByte
-      i += 1
-    }
-    result
+  def readBigDecimal(): BigDecimal = matchNumericString(BigDecimal(_, options.mathContext))
+
+  override def readTimestamp(): Long = options.dateFormat match {
+    case JsonDateFormat.EpochMillis => readLong()
+    case JsonDateFormat.IsoInstant => IsoInstant.parse(readString())
+  }
+
+  def readBinary(): Array[Byte] = options.binaryFormat match {
+    case JsonBinaryFormat.ByteArray =>
+      val builder = new mutable.ArrayBuilder.ofByte
+      val li = readList()
+      while (li.hasNext) {
+        builder += li.nextElement().readByte()
+      }
+      builder.result()
+    case JsonBinaryFormat.HexString =>
+      val hex = checkedValue[String](JsonType.string)
+      val result = new Array[Byte](hex.length / 2)
+      var i = 0
```

Comment:
`(0 until result.length).map(...` should not be much slower and you could even go all the way `.toArray`

Reply:
I want this as raw and fast as possible.

Reply:
Benchmark it then :smiley:

Reply:
I will. But on the first sight your version should be much slower because not only will it create intermediate collection but also **every byte will be boxed** in that collection.

Reply:
Introducing `foreach` was relatively harmless but `map` + `toArray` cut off more than half of benchmark score. Ultimately I made a tail-recursive loop.